src/Data/LargeHashable/Intern.hs

```diff
@@ -6,67 +6,145 @@
+    , hu_updateUChar :: {-# NOUNPACK #-} !(CUChar -> IO ())    -- Word8
+    , hu_updateShort :: {-# NOUNPACK #-} !(CShort -> IO ())    -- Int16
+    , hu_updateUShort :: {-# NOUNPACK #-} !(CUShort -> IO ())  -- Word16
+    , hu_updateInt :: {-# NOUNPACK #-} !(CInt -> IO ())        -- Int32
+    , hu_updateUInt :: {-# NOUNPACK #-} !(CUInt -> IO ())      -- Word32
+    , hu_updateLong :: {-# NOUNPACK #-} !(CLong -> IO ())      -- Int64
+    , hu_updateULong :: {-# NOUNPACK #-} !(CULong -> IO ())    -- Word64
     }
 
 -- | The interface for a hashing algorithm. The interface contains a simple run
 -- function, which is used to update the hash with all values needed, and the
 -- outputs the resulting hash.
-data HashAlgorithm
+data HashAlgorithm h
     = HashAlgorithm
-    { ha_run :: (HashUpdates -> IO ()) -> IO Hash
+    { ha_run :: {-# NOUNPACK #-} !((HashUpdates -> IO ()) -> IO h)
+    , ha_xor :: {-# NOUNPACK #-} !(h -> h -> h)
+    , ha_updateHash :: {-# NOUNPACK #-} !(HashUpdates -> h -> IO ())
+    }
+
+data LHEnv
+    = LHEnv
+    { lh_updates :: {-# NOUNPACK #-} !HashUpdates
+    , lh_updateXorHash :: {-# NOUNPACK #-} !([LH ()] -> IO ())
     }
 
 -- | The 'LH' monad (`LH` stands for "large hash") is used in the definition of
 -- hashing functions for arbitrary data types.
-newtype LH a = LH { _unLH :: ReaderT HashUpdates IO a }
-    deriving (Functor, Applicative, Monad)
+newtype LH a = LH (LHEnv -> IO a)
+
+{-# INLINE lhFmap #-}
+lhFmap :: (a -> b) -> LH a -> LH b
+lhFmap f (LH x) =
+    LH $ \env ->
+    do y <- x env
+       return (f y)
+
+{-# INLINE lhReturn #-}
+lhReturn :: a -> LH a
+lhReturn x = LH $ \_env -> return x
+
+{-# INLINE lhApp #-}
+lhApp :: LH (a -> b) -> LH a -> LH b
+lhApp (LH f) (LH x) =
+    LH $ \env -> f env <*> x env
 
+{-# INLINE lhBind #-}
+lhBind :: LH a -> (a -> LH b) -> LH b
+lhBind (LH x) f =
+    LH $ \env ->
+    do y <- x env
+       let (LH g) = f y
+       g env
+
+{-# INLINE lhBind' #-}
+lhBind' :: LH a -> LH b -> LH b
+lhBind' (LH x) (LH y) =
+    LH $ \env ->
+    do _ <- x env
+       y env
+
+instance Functor LH where
+    fmap = lhFmap
+
+instance Applicative LH where
+    pure = lhReturn
+    (<*>) = lhApp
+
+instance Monad LH where
+    return = lhReturn
+    (>>=) = lhBind
+    (>>) = lhBind'
+
+{-# INLINE hashUpdates #-}
 hashUpdates :: LH HashUpdates
-hashUpdates = LH ask
+hashUpdates =
+    LH $ \env -> return (lh_updates env)
+
+{-# INLINE getUpdateXorHash #-}
+getUpdateXorHash :: LH ([LH ()] -> IO ())
+getUpdateXorHash =
+    LH $ \env -> return (lh_updateXorHash env)
 
 -- | Perform an 'IO' action in the 'LH' monad. Use with care, do not perform
 -- arbitrary 'IO' operation with this function! Only use it for calling
 -- functions of the 'HashUpdates' datatype.
+{-# INLINE ioInLH #-}
 ioInLH :: IO a -> LH a
-ioInLH = LH . liftIO
+ioInLH io =
+    LH $ \_env -> io
 
 -- | Runs a 'LH' computation and returns the resulting hash.
 {-# NOINLINE runLH #-}
```

Comment:
It's needed because of the unsafePerformIO, see https://hackage.haskell.org/package/base-4.9.0.0/docs/System-IO-Unsafe.html